### ISSUE TITLE
Provide self containing binary

### ISSUE THREAD

**Issue by abergmeier** [microsoft/pyright#1641]
**Is your feature request related to a problem? Please describe.**
Especially for CI and automation it would be easiest to have a self containing binary, which could be dropped anywhere and be executed.

**Describe the solution you'd like**
It would be good to have a binary, which contains everything necessary to run pyright binary. Either a shell script for extracting into cache directory first and then running or some other means of packaging.

**Additional context**
Installing npm often is not an option in an environment purely using Python code.

**Comment by jakebailey** [comment]
This was previously presented as #1414; at the time we were worried about the package size and the maintainability (given we'd depend on `vercel/pkg` to build binaries, and it was dead). Since then, I've seen packages on PyPI that are much more heavily used and are larger (like tensorflow, which is hundreds of megs and certainly more often downloaded), and `vercel/pkg` has regained at least some maintenance again. I think it might be worth prototyping something again to see if it'd work as expected and what the pitfalls of the one-binary would be.

Note that for "CI and automation", it'd greatly surprise me if node was difficult to obtain. All CI services I know of have directives to get a specific version of node, and can run arbitrary commands to do things. Many projects successfully make pyright work either with manual installs or with precommit.

**Comment by erictraut** [comment]
Please refer to [this documentation](https://github.com/microsoft/pyright/blob/master/docs/ci-integration.md). It shows how to integrate pyright into CI using a pre-commit hook and a shell script.

As Jake said, we looked previously at publishing pyright in some other packaged form, but there are some significant downsides of doing so. I don't think the circumstances or assumptions have changed in the past few months, so it's unlikely that we'd come to a different decision now.

**Comment by abergmeier** [comment]
> Please refer to this documentation. It shows how to integrate pyright into CI using a pre-commit hook and a shell script.

For CI, I personally, would rather package pyright into a container instead. This way the complexity of maintaining npm is at least limited mostly to building the container.

> As Jake said, we looked previously at publishing pyright in some other packaged form, but there are some significant downsides of doing so. I don't think the circumstances or assumptions have changed in the past few months, so it's unlikely that we'd come to a different decision now.

Thanks for thinking about this. I do understand that currently there may be other priorities.

**Comment by jakebailey** [comment]
Docker isn't such a bad idea and is much easier to automate than PyPI. I've also been contemplating spending a little time to make a GitHub actions wrapper (with the fancy diagnostic handling). What CI are you using, out of curiosity?

**Comment by abergmeier** [comment]
> What CI are you using, out of curiosity?

We are debating between GitHub Actions and Google Cloud Build.

**Comment by jakebailey** [comment]
Thanks for the info. You may be interested in what I added to typeshed to test stubs with pyright: https://github.com/python/typeshed/blob/fc2933ca648a2227457c3bd1490d34677405d065/.github/workflows/tests.yml#L89

The wrapper script is redundant and mainly for local dev users to try and run things; pyright can be run directly of course:

```yml
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v2
        with:
          node-version: '14'
      - run: npx pyright
```

**Comment by kumaraditya303** [comment]
@jakebailey  @erictraut  I would suggest to have a look at [pyre-check](https://github.com/facebook/pyre-check), python type checker from facebook which does the same thing as my issue #1414 and this issue, the wheel of it are also uploaded to PyPI  https://pypi.org/project/pyre-check/#files so that users can just drop it and use it. from python as a pip package.

The `pkg` packages seems more maintained now so it is worth exploring it. In case, the maintainers agree then  I can refactor my original PR #1415 for it

**Comment by erictraut** [comment]
@abergmeier, have the above suggestions addressed your needs?

As I said above, we looked at providing a self-contained binary a couple of months back, but after much deliberation and exploration we decided against it. I don't think anything has changed in the interim that would change the outcome of that decision.

**Comment by erictraut** [comment]
@bergmeier, since we haven't heard back from you, I'm going to close this request. Feel free to reply when you have a chance.
